docs: add k3s pack doc (#1792)
spectrocloud/librarium: docs/docs-content/integrations/k3s.md

---
sidebar_label: 'K3s'
title: 'K3s'
description: 'K3s pack in Palette Edge'
hide_table_of_contents: true
type: "integration"
category: ['kubernetes', 'amd64', 'fips']
sidebar_class_name: "hide-from-sidebar"
logoUrl: 'https://registry.spectrocloud.com/v1/edge-k3s/blobs/sha256:5b7f8a80330d6938056848905238efbbd085e2f0f86fd03d7be10ccdabf791c7?type=image/png'
tags: ["packs", "kubernetes", "k3s", "edge"]
---

> K3s is a lightweight distribution of Kubernetes, specifically designed for resource-constrained environments and IoT applications. Unlike standard Kubernetes with multiple binaries for its various components, K3s is packaged as a single binary with a lightweight storage backend and minimal external dependencies. For more information about K3s, refer to [K3s documentation](https://docs.k3s.io/). 

Reviewer: ```suggestion
K3s is a lightweight distribution of Kubernetes, specifically designed for resource-constrained environments and Internet of Things (IoT) applications. Unlike standard Kubernetes with multiple binaries for its various components, K3s is packaged as a single binary with a lightweight storage backend and minimal external dependencies. For more information about K3s, refer to [K3s documentation](https://docs.k3s.io/). 
```

## Versions Supported

> <Tabs queryString="versions">

Reviewer: Looking at the Packs reporter, I see active versions in prod. We also need to specify these versions and created a tab for deprecated stating what is the cutoff.

![CleanShot 2023-11-15 at 09 48 54](https://github.com/spectrocloud/librarium/assets/29551334/acdd78e4-3a23-4ea1-965e-f262d07e1052)

Author: yeah, I was going to make all these versions. I just wanted to make sure I had the right content before I coped them to all the other version tabs.

<TabItem label="1.27.X" value="k3s_1.27">

### Prerequisites

- An edge device with AMD64(x86_64) processor architecture or a Palette Virtual Cluster. 
- A minimum of 2 CPU cores and 1 GB memory. 

### Parameters

> Since you can deploy both Palette Virtual Clusters and Edge clusters using K3s, you have different configuration options depending on the type of cluster.

Reviewer: @lennessyy could this be presented as a tab instead of two headings? That might simplify things for the reader.

Author: I tried to make them tabs, but it was giving me build errors and i thought we cannot do tabs in tabs since it's already in the version tab. But maybe I just did the syntax wrong and I'll try again

> Since you can deploy both Palette Virtual Clusters and Edge clusters using K3s, you have different configuration options depending on the type of cluster.

Reviewer 2: ```suggestion
Since you can deploy both virtual clusters and Edge clusters using K3s, you have different configuration options depending on the cluster type.
```

Reviewer 2: The first occurrence within a doc:  Palette Virtual Cluster

Subsequent occurrences within the same doc: virtual cluster

#### Edge Cluster

> The following is a list of important parameters when you deploy an Edge cluster using K3s:

Reviewer 2: Are these required parameters, or are they good to know if users want to use them? If required, you might want to change "important" to "required". 

If good to know about, you could potentially remove the sentence altogether as in other refactored packs docs.

|`kube-controller-manager-arg` | This parameter describes extra arguments for the Kubernetes Controller Manager, such as enabling certain feature gates and setting profiling. |
| `kubelet-arg` |  This parameter contains extra arguments for Kubelet during node registration, such as setting feature gates, protecting kernel defaults, and disabling the read-only port. |
| `pack.palette.config.oidc.identityProvider` | Dynamically enabled OpenID Connect (OIDC) Identity Provider (IDP) setting based on your UI selection when you add the K3s pack to your profile. This parameter appears in the YAML file after you make a selection. Refer to [Configure OIDC Identity Provider](#configure-custom-oidc). |

You can add cloud-init stages, which allow you to customize your instances declaratively. The cloud-init stages are exposed by [Kairos](https://kairos.io/docs/architecture/cloud-init/), an open source project. For more information, check out the [Cloud Init Stages](../clusters/edge/edge-configuration/cloud-init.md) reference.

#### Palette Virtual Cluster

> Since the Palette Virtual Cluster is a virtual cluster that you are setting up inside another Kubernetes cluster, you can configure its pods and services differently than the host cluster. The kubeadm file you get includes parameters that offer you a higher degree of customization. 

Reviewer 2: ```suggestion
Since you are setting up a virtual cluster inside another Kubernetes cluster, you can configure its pods and services differently than the host cluster. The Kubeadm configuration file you get includes parameters that offer you a higher degree of customization. 
```

Reviewer 2: "...you get" - is the Kubeadm file generated? If so, this could say "The Kubeadm configuration file that gets generated includes parameters..." or "The generated Kubeadm configuration file..."

> The following is a list of important parameters when you deploy a Palette Virtual Cluster cluster using K3s:

Reviewer 2: ```suggestion
The following is a list of important parameters when you deploy a virtual cluster using K3s:
```

|**Parameter**|**Description** |
|-------------|----------------|
|`enableHA`| Determines whether the control plane is deployed in high availability mode. If you set this parameter to true, make sure to adjust the number of replicas and use an external datastore. |
|`defaultImageRegistry` | Specifies the default registry from which images are pulled. The value of this parameter will be prepended to all deployed virtual cluster images. If an image has already been deployed as part of the virtual cluster, the deployed images within the virtual cluster will not be rewritten.|
|`sync` | Specifies which Kubernetes resources are synced between the virtual and host clusters. |
|`storage` | Specifies storage settings such as persistence and PVC size. By default, storage of the virtual cluster uses the same storage class as the host cluster, but you can also optionally specify a different storage class. |

> |`ingress` | Configures the ingress resource that allows you to access the virtual cluster |

Reviewer 2: ```suggestion
|`ingress` | Configures the ingress resource that allows you to access the virtual cluster. |
```

### Usage

> K3s is available for Edge host deployments as well as Palette Virtual Clusters that you can create from cluster groups. Refer to the [Create an Infrastructure Profile](../profiles/cluster-profiles/create-cluster-profiles/create-infrastructure-profile.md) guide and the [Create and Manage Cluster Groups](https://docs.spectrocloud.com/clusters/cluster-groups/create-cluster-group#palette-virtual-cluster-configuration) guide for more information. 

Reviewer 2: ```suggestion
K3s is available for Edge host deployments and virtual clusters that you can create from cluster groups. Refer to the [Create an Infrastructure Profile](../profiles/cluster-profiles/create-cluster-profiles/create-infrastructure-profile.md) guide and the [Create and Manage Cluster Groups](https://docs.spectrocloud.com/clusters/cluster-groups/create-cluster-group#palette-virtual-cluster-configuration) guide for more information. 
```

Reviewer 2: In the second link in line 58, remove "https://docs.spectrocloud.com" - you might also have to adjust the relative path.

:::info

> In order to use K3s as part of an Edge deployment, you need to go through the EdgeForge process and specify K3s as your intended Kubernetes distribution when you build your OS image. For more information, refer to the [EdgeForge Workflow](https://docs.spectrocloud.com/clusters/edge/edgeforge-workflow/) guide. 

Reviewer 2: ```suggestion
To use K3s as part of an Edge deployment, you need to go through the EdgeForge process and specify K3s as your intended Kubernetes distribution when you build your OS image. For more information, refer to the [EdgeForge Workflow](https://docs.spectrocloud.com/clusters/edge/edgeforge-workflow/) guide. 
```

Reviewer 2: Remove "https://docs.spectrocloud.com" and adjust the relative path.

:::

> #### Configure OIDC Identity Provider for Edge Deployments

Reviewer: ```suggestion
#### Configure OIDC Identity Provider for Edge
```
Let's try to keep titles short.

What about Cluster Group? Could we mention that and link to the other [guide](https://docs.spectrocloud.com/clusters/palette-virtual-clusters/configure-oidc-virtual-cluster/#validate)?

Author: I linked this one in the next heading, though I mentioned Palette Virtual Cluster instead of cluster group since that is the wording we used in the other guide.

> You can modify the kubeadm file to configure your Edge cluster to use an OIDC Identity Provider (IDP) for authentication. 

Reviewer 2: ```suggestion
You can modify the kubeadm file to configure your Edge cluster to use an OpenID Connect (OIDC) Identity Provider (IDP) for authentication. 
```

> You can use a custom third-party IDP such as Okta or you can use Palette as your IDP. 

Reviewer 2: ```suggestion
You can use a custom third-party IDP, such as Okta, or use Palette as your IDP. 
```

> When you add the K3s pack to a cluster profile, Palette displays the OIDC IDP options listed below:

Reviewer 2: ```suggestion
When you add the K3s pack to a cluster profile, Palette displays OIDC IDP options, which you can configure in the Kubernetes layer. To learn about these options, refer to the [Palette eXtended Kubernetes - Edge](/integrations/kubernetes-edge/#configure-oidc-identity-provider) section.
```

Reviewer 2: Instead of duplicating content, we typically document it in one place and link to it because it makes content management easier - this allows us to make updates in one place. All of the OIDC information that follows is documented in the Kubernetes pack docs (PXK and PXK-E). 

If the user _must_ configure Custom, you could tell them that and refer them to existing documentation.  

Note that the link above is not a relative path, and I'm not sure that's where you want to take your reader for Palette Virtual Cluster. OIDC is also documented here for Palette Virtual Clusters: https://docs.spectrocloud.com/clusters/palette-virtual-clusters/configure-oidc-virtual-cluster

Author: For this one, it is true that it is duplicating content, but i feel like users who are using this pack wouldn't want to go to another pack for information. I think this duplication is okay for now. Maybe in the future we can move both content to a separate page to something like "Configuring OIDC for Edge" under the edge documentation and link to that page then

   ```yaml
   kubeadmconfig:
    clientConfig:
      oidc-issuer-url: "<OIDC-ISSUER-URL>"
      oidc-client-id: "<OIDC-CLIENT-ID>"
      oidc-client-secret: "<OIDC-CLIENT-SECRET>"
      oidc-extra-scope: profile,email,openid
   ```

After you have configured the IDP for authentication, you can proceed to create role bindings to configure authorization in your cluster. Refer to [Create Role Bindings](../clusters/cluster-management/cluster-rbac.md#create-role-bindings) for more guidance. 

#### Configure OIDC Identity Provider for Palette Virtual Clusters

If you are using K3s in a Palette Virtual Cluster inside of a cluster group, you can also configure OIDC for your cluster. Refer to [Configure OIDC for a Virtual Cluster](../clusters/palette-virtual-clusters/configure-oidc-virtual-cluster.md) for more guidance.

#### Add a Certificate for Reverse Proxy

You can use a reverse proxy with a K3s Kubernetes cluster. The reverse proxy allows you to connect to the cluster API of a Palette-managed Kubernetes cluster in private networks or clusters configured with private API endpoints. For more information, refer to the [Spectro Proxy](frp.md) guide.

> ### Terraform

Reviewer: We keep the Terraform section outside of the tabs.

> 

Reviewer: Let's add a sentence here. Take a peek at other packs for examples.

|`kube-controller-manager-arg` | This parameter describes extra arguments for the Kubernetes Controller Manager, such as enabling certain feature gates and setting profiling. |
| `kubelet-arg` |  This parameter contains extra arguments for Kubelet during node registration, such as setting feature gates, protecting kernel defaults, and disabling the read-only port. |
| `pack.palette.config.oidc.identityProvider` | Dynamically enabled OpenID Connect (OIDC) Identity Provider (IDP) setting based on your UI selection when you add the K3s pack to your profile. This parameter appears in the YAML file after you make a selection. Refer to [Configure OIDC Identity Provider](#configure-custom-oidc). |

You can add cloud-init stages, which allow you to customize your instances declaratively. The cloud-init stages are exposed by [Kairos](https://kairos.io/docs/architecture/cloud-init/), an open source project. For more information, check out the [Cloud Init Stages](../clusters/edge/edge-configuration/cloud-init.md) reference.
</TabItem>

<TabItem label="Palette Virtual Cluster" value="palette-virtual-cluster">

> Since you are setting up a virtual cluster inside another Kubernetes cluster, you can configure its pods and services differently than the host cluster. The default Kubeadm configuration file you get includes parameters that offer you a higher degree of customization.

Author: ```suggestion
Since you are setting up a virtual cluster inside another Kubernetes cluster, you can configure its pods and services differently than the host cluster. The default kubeadm configuration file you get includes parameters that offer you a higher degree of customization.
```

:::info
In order to use K3s as part of an Edge deployment, you need to go through the EdgeForge process and specify K3s as your intended Kubernetes distribution when you build your OS image. For more information, refer to the [EdgeForge Workflow](../clusters/edge/edgeforge-workflow/) guide. 
:::

#### Configure OIDC Identity Provider for Edge

You can modify the configuration file to configure your Edge cluster to use an OpenID Connect (OIDC) Identity Provider (IDP) for authentication. 
You can use a custom third-party IDP, such as Okta, or use Palette as your IDP. 

When you add the K3s pack to a cluster profile, Palette displays the OIDC IDP options listed below:

- **None**: This setting does not require OIDC configuration for the cluster. It displays in the YAML file as `noauth`. 

- **Custom**: This is the default setting and does not require OIDC configuration. However, if desired, it allows you to specify a third-party OIDC provider by configuring OIDC statements in the YAML file as described in [Configure Custom OIDC](kubernetes-edge.md#configure-custom-oidc). This setting displays in the YAML file as `none`.

> - **Palette**: This setting makes Palette the IDP. Any user with a Palette account in the tenant and the proper permissions to view and access the project's resources is able to use kubectl CLI to access cluster. This setting displays in the YAML file as `palette`. Selecting **Palette** in this setting and creating role bindings to configure authorization are all you need to do to enable OIDC for your cluster. You do not need to provide extra parameters such as `oidc-issuer-url` as you need to when you configure a custom OIDC provider. 

Reviewer 2: ```suggestion
- **Palette**: This setting makes Palette the IDP. Any user with a Palette account in the tenant and the proper permissions to view and access the project's resources is able to use kubectl CLI to access cluster. This setting displays in the YAML file as `palette`. When you select **Palette**, all you have to do to enable OIDC for your cluster is create role bindings to configure authorization. You do not need to provide extra parameters such as `oidc-issuer-url` as you need to when you configure a custom OIDC provider. 
```

Reviewer 2: This was a good bit of information to add. I slightly reworded because I had to reread the sentence a couple times to get the meaning at first.
